### ISSUE TITLE
Unflake TestCancelUpload

### ISSUE THREAD

**Issue by jadekler** [googleapis/google-api-go-client#395]
https://sponge.corp.google.com/target?id=fbae2ebe-08f8-4067-b1e2-6d8fb41fe2fd&target=cloud-go-libraries%2Fgoogle-api-go-client%2Fpresubmit%2Fgo111&searchFor=

```
--- FAIL: TestCancelUpload (0.00s)
panic: Ran out of events, but got a request [recovered]
	panic: Ran out of events, but got a request

goroutine 59 [running]:
testing.tRunner.func1(0xc00012d100)
	/usr/local/go/src/testing/testing.go:792 +0x6a7
panic(0xa61d40, 0xbb53f0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
google.golang.org/api/gensupport.(*interruptibleTransport).RoundTrip(0xc00009e780, 0xc00012db00, 0xc00009e780, 0x0, 0x0)
	/root/go/src/google.golang.org/api/gensupport/resumable_test.go:69 +0x9fe
net/http.send(0xc00012db00, 0xbb77a0, 0xc00009e780, 0x0, 0x0, 0x0, 0xc00018f5b0, 0x46d8ef, 0xc0001d41f0, 0x0)
	/usr/local/go/src/net/http/client.go:250 +0x305
net/http.(*Client).send(0xc0001e2bd0, 0xc00012db00, 0x0, 0x0, 0x0, 0xc0001d41f0, 0x0, 0x1, 0xc00013e6a9)
	/usr/local/go/src/net/http/client.go:174 +0x1cb
net/http.(*Client).do(0xc0001e2bd0, 0xc00012db00, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/client.go:641 +0x53b
net/http.(*Client).Do(0xc0001e2bd0, 0xc00012db00, 0x10, 0xc00008da50, 0x0)
	/usr/local/go/src/net/http/client.go:509 +0x43
google.golang.org/api/gensupport.send(0xbbc4a0, 0xc00009e7c0, 0xc0001e2bd0, 0xc00012da00, 0x0, 0x0, 0xc000198a70)
	/root/go/src/google.golang.org/api/gensupport/send.go:67 +0x163
google.golang.org/api/gensupport.SendRequest(0xbbc4a0, 0xc00009e7c0, 0xc0001e2bd0, 0xc00012da00, 0x3, 0xc0000cc580, 0x11)
	/root/go/src/google.golang.org/api/gensupport/send.go:52 +0x23d
google.golang.org/api/gensupport.(*ResumableUpload).doUploadRequest(0xc0001f66c0, 0xbbc4a0, 0xc00009e7c0, 0xbb7440, 0xc0001e3140, 0x10e, 0x1e, 0xf72a01, 0xc00018fb00, 0x475819, ...)
	/root/go/src/google.golang.org/api/gensupport/resumable.go:101 +0x361
google.golang.org/api/gensupport.(*ResumableUpload).transferChunk(0xc0001f66c0, 0xbbc4a0, 0xc00009e7c0, 0x1, 0x1, 0x0)
	/root/go/src/google.golang.org/api/gensupport/resumable.go:133 +0x13f
google.golang.org/api/gensupport.(*ResumableUpload).Upload(0xc0001f66c0, 0xbbc4a0, 0xc00009e7c0, 0xc0001fe240, 0x0, 0x5a)
	/root/go/src/google.golang.org/api/gensupport/resumable.go:214 +0x1e1
google.golang.org/api/gensupport.TestCancelUpload(0xc00012d100)
	/root/go/src/google.golang.org/api/gensupport/resumable_test.go:286 +0x835
testing.tRunner(0xc00012d100, 0xb358a0)
	/usr/local/go/src/testing/testing.go:827 +0x163
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:878 +0x65a
FAIL	google.golang.org/api/gensupport	0.139s
?   	google.golang.org/api/gmail/v1	[no test files]
```

**Comment by jadekler** [comment]
@odeke-em I'm pretty out of cycles right now. Do you have time to take a peek at this?

**Comment by odeke-em** [comment]
Hey Jean,
  Thank you for pinging me! Absolutely I'll work on this.

Kind regards,
Emmanuel

**Comment by odeke-em** [comment]
Interestingly on the second run of `go test gensupport` I was able to get a similar looking crash
```shell
$ go test
panic: Ran out of events, but got a request

goroutine 75 [running]:
google.golang.org/api/gensupport.(*interruptibleTransport).RoundTrip(0xc0000f8c40, 0xc000812c00, 0xc0000f8c40, 0x0, 0x0)
	/Users/emmanuelodeke/go/src/google.golang.org/api/gensupport/resumable_test.go:69 +0x6c6
net/http.send(0xc000812c00, 0x15d8da0, 0xc0000f8c40, 0x0, 0x0, 0x0, 0xc000828058, 0x0, 0x1, 0x0)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/net/http/client.go:250 +0x435
net/http.(*Client).send(0xc00014e960, 0xc000812c00, 0x0, 0x0, 0x0, 0xc000828058, 0x0, 0x1, 0x0)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/net/http/client.go:174 +0xfa
net/http.(*Client).do(0xc00014e960, 0xc000812c00, 0x0, 0x0, 0x0)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/net/http/client.go:641 +0x3ce
net/http.(*Client).Do(...)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/net/http/client.go:509
google.golang.org/api/gensupport.send(0x15e25a0, 0xc0000ba008, 0xc00014e960, 0xc000812b00, 0x192c300, 0xc00006dc00, 0xdaee5243)
	/Users/emmanuelodeke/go/src/google.golang.org/api/gensupport/send.go:67 +0xef
google.golang.org/api/gensupport.SendRequest(0x15e25a0, 0xc0000ba008, 0xc00014e960, 0xc000812b00, 0xc000822ff0, 0xc0007f6380, 0xc)
	/Users/emmanuelodeke/go/src/google.golang.org/api/gensupport/send.go:52 +0x193
google.golang.org/api/gensupport.(*ResumableUpload).doUploadRequest(0xc0000b4780, 0x15e25a0, 0xc0000ba008, 0x15d8a40, 0xc0007f0660, 0x0, 0x5a, 0xc0000b4900, 0xc00008f708, 0xc00008f7b8, ...)
	/Users/emmanuelodeke/go/src/google.golang.org/api/gensupport/resumable.go:101 +0x54f
google.golang.org/api/gensupport.(*ResumableUpload).transferChunk(0xc0000b4780, 0x15e25a0, 0xc0000ba008, 0x1, 0x1, 0x0)
	/Users/emmanuelodeke/go/src/google.golang.org/api/gensupport/resumable.go:133 +0xc9
google.golang.org/api/gensupport.(*ResumableUpload).Upload(0xc0000b4780, 0x15e25a0, 0xc0000ba008, 0xc0009b9f70, 0x10f9f39, 0x1555808)
	/Users/emmanuelodeke/go/src/google.golang.org/api/gensupport/resumable.go:214 +0x196
google.golang.org/api/gensupport.TestRetry_Bounded.func5(0xc0000b4780, 0xc0001aa400, 0xc00012a070)
	/Users/emmanuelodeke/go/src/google.golang.org/api/gensupport/resumable_test.go:345 +0x47
created by google.golang.org/api/gensupport.TestRetry_Bounded
	/Users/emmanuelodeke/go/src/google.golang.org/api/gensupport/resumable_test.go:344 +0x3e3
exit status 2
FAIL	google.golang.org/api/gensupport	2.030s
```
but on a different test.

**Comment by odeke-em** [comment]
I just spent time examining how this could happen and the theory here is that when Cancel was invoked in the progress reporter, due to a timing issue, net/http's client processing requests in a goroutine for persistConn.roundTrip or persistConn.writeLoop got a write before the cancellation on the the ctxDoneChan was received and then made that request i.e. the cancellation occured concurrently in the midst of a request.

I can't reproduce the failure physically but perhaps we could improve how to catch it by printing out the state of the world in the panic to validate assumptions because a dump of active goroutines would show us what's happening, and this we could do by a helper routine that'll invoke `runtime.Stack(buf, true)` and also we can print out the stray request by `httputil.DumpRequest(req, true)`

To remedy this issue for good, perhaps what we can do is:
a) add some a small sleep in the progressRecorder after cancelFunc() is invoked e.g.
100ms with a comment such as
```go
        pr := progressRecorder{f: func() {
                numUpdates++
                if numUpdates >= 2 {
                        cancelFunc()
                        // Give cancellation time to propagate to avoid racing against in-flight requests
                         // as seen in https://github.com/googleapis/google-api-go-client/issues/395
                         time.Sleep(100 * time.Millisecond)
                }
        }}
```

but unfortunately my suggested remedy is reactionary and not fully data driven!

I think what we can do is add the part that prints out the state of the world and then on any flake,
we'll also suggest they come back and re-open this issue.

**Comment by escholtz** [comment]
Looks like this was fixed in #400?
